aip/general/0129.md

```diff
@@ -0,0 +1,119 @@
+---
+id: 129
+state: draft
+created: 2023-06-21
+placement:
+  category: resource-design
+  order: 65
+---
+
+# Server-Modified Values and Defaults
+
+Server-modified and default values tend to cause problems for
+[state-driven clients](state-driven-clients). These clients are often unable to
+tell when their desired state matches the current state for these fields, as the
+rules by which a server may modify and return values are not public and
+repeatable. This AIP aims to clarify what behaviors can be expected from
+services and describing how they can be repeated on the client-side.
+
+## Guidance
+
+### Effective Values
+There are instances where a service will allocate, generate, or calculate a
+value if the client chooses not to specify one. For example: a client may
+create a virtual machine without specifying a static IP address for the virtual
+machine to be available on. Such a scenario is opting into dynamic IP address
+allocation.
+
+Some examples of these types of fields are ones that are:
+
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with an effective value ***must*** be expressed as two fields in
+the API:
+* a mutable field that can be optionally set by the user and **must not** be
+modified by the service
+* an `OUTPUT_ONLY` field that records the effective value decided on by the
+service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string effective_ip_address = 5 [
+    (google.api.field_behavior) = OUTPUT_ONLY
+  ];
+}
+```
+
+#### Naming
+
+Effective values *should* be named by prefixing `effective_` to the mutable
+field's name. Consistent naming is important for identifying standard behavior
+across APIs and fields.
+
+### Client-Settable Fields and Normalization
+
+For client-settable fields, the value in response from the service **must** be
+the same as provided by the create or update request. For string fields this
+means returning the unchanged character sequence, with two exceptions: 
+
+* When a string field has a data type annotation, a normalized string that
+represents the given value may be returned.
+* When a string field has a default value annotation, that default value may be
+returned when the given value is empty or missing.
+Normalizations are important to allow services to store and return values in a
```

Comment:
move "normalization" to it's own section? it's a whole topic in it's own right.

Reply:
Moved to its own section and moved most of this into a rationale section